src/Featurizers/Components/DocumentStatisticsEstimator.h

```diff
@@ -88,26 +106,110 @@ class DocumentStatisticsAnnotationData {
+    DocumentStatisticsAnnotationData(FrequencyAndIndexMap termFrequencyAndIndex, std::uint32_t totalNumDocuments);
     ~DocumentStatisticsAnnotationData(void) = default;
 
     FEATURIZER_MOVE_CONSTRUCTOR_ONLY(DocumentStatisticsAnnotationData);
 };
 
+namespace {
+
+inline DocumentStatisticsAnnotationData::FrequencyMap PruneTermFreqMap(DocumentStatisticsAnnotationData::FrequencyMap const & prunedTermFreqByDf, 
+                                                                       std::uint32_t maxFeatures) {
+    std::vector<std::uint32_t> frequencyVector;
+
+    std::transform(prunedTermFreqByDf.begin(), 
+                   prunedTermFreqByDf.end(), 
+                   std::back_inserter(frequencyVector), 
+                   [](auto const & kv) { 
+                       return kv.second; 
+                   });
+
+    //partial sort, ave O(n)
+    std::nth_element(frequencyVector.begin(), 
+                     frequencyVector.begin() + static_cast<int>(maxFeatures) - 1, 
+                     frequencyVector.end(), 
+                     std::greater<std::uint32_t>());
+
+    std::uint32_t targetValue = frequencyVector[maxFeatures - 1];
+    long long remainingTargetValue = std::count(frequencyVector.begin(), 
+                                                frequencyVector.begin() + static_cast<int>(maxFeatures),
+                                                targetValue);
+
+    DocumentStatisticsAnnotationData::FrequencyMap prunedTermFreq;
+    for (auto const & termFrequencyPair : prunedTermFreqByDf) {
+
+        if (termFrequencyPair.second > targetValue)
+            prunedTermFreq.insert(termFrequencyPair);
+
+        if (termFrequencyPair.second == targetValue && remainingTargetValue > 0) {
+            prunedTermFreq.insert(termFrequencyPair);
+            --remainingTargetValue;
+        }
+    }
+
+    return prunedTermFreq;
+}
+
+inline DocumentStatisticsAnnotationData::FrequencyMap PruneTermFreqMap(DocumentStatisticsAnnotationData::FrequencyMap const & termFrequency, 
+                                                                       std::float_t minDf, 
+                                                                       std::float_t maxDf, 
+                                                                       std::float_t totalNumDocumentsFloat) {
+    if (!(minDf > 0.0f || maxDf < 1.0f))
+        return termFrequency;
+
+    DocumentStatisticsAnnotationData::FrequencyMap prunedTermFreq;
+    for (auto const & termFrequencyPair : termFrequency) {
+        std::float_t const freq = termFrequencyPair.second / totalNumDocumentsFloat;
+
+        if (freq >= minDf && freq <= maxDf) 
+            prunedTermFreq.insert(termFrequencyPair);
+    }
+
+    return prunedTermFreq;
+}
+
+inline DocumentStatisticsAnnotationData::FrequencyMap PruneTermFreqMap(DocumentStatisticsAnnotationData::FrequencyMap prunedTermFreqByDf, 
```

Comment:
This method is very close to the definition above. Are they both necessary? If so, move them so they are defined next to each other. Also, add a comment explaining why they are both necessary.

Reply:
Agree, I believe it's better to do in one function